[DRAFT] DOCS:Next.js (APP ROUTER)

PedroAmorimG opened the pull request:
Arrumei o que havia sido falado e adicionei a parte de Server Action, Caching e Optimizing.

PedroAmorimG committed:
Correção, Server Actions, caching e optimizações

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+!!!
+O uso de um `input` com atributo `hidden` pode ser útil quando precisar de passar alguma informação para a `server action` que não será fornecida pelo usuário.
+!!!
```
```suggestion
A função `revalidatePath` possui uma interação interessante com o uso de `Server Actions`. Quando juntas elas podem criar certo grau de interatividade na página.
!!!

!!!
O uso de um `input` com atributo `hidden` pode ser útil quando precisar passar alguma informação para a `server action` que não será fornecida pelo usuário.
!!!
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+### Server Components
+
+Os `server components` são utilizados para renderização de UI no lado do servidor e podem sofrer `cache`.
```
```suggestion
Os `server components` são utilizados para renderização de UI no lado do servidor e podem ser beneficiados com `caching`.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+Se o cache for mal feito, a página pode conter erros como dados desatualizados.
+!!!
+
+Por padrão, o next coloca pedaços estáticos de páginas e resultados de `fetch` no `cache`, contudo, é possível manipular se algo irá para `cache` ou não, quanto tempo permancerá em `cache` e até atualizar o `cache` após certo evento. Para isso, poderão ser utilizados diferentes métodos como `cache()`, `revalidatePath()` e configurações de rota, por exemplo.
```
```suggestion
Por padrão, o next coloca pedaços estáticos de páginas e resultados de `fetch` no `cache`. Contudo, é possível manipular se algo irá para `cache` ou não, quanto tempo permancerá em `cache` e até atualizar o `cache` após certo evento. Para isso, poderão ser utilizados diferentes métodos como `cache()`, `revalidatePath()` e configurações de rota, por exemplo.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+#### Fetch
+
+Resultados recebidos pelo `fetch`, quando utilizado `server side` são memorizados até que o componente seja renderizado, por padrão, mas podem ser ajustados passando mais parâmetros para a função.
```
```suggestion
Resultados recebidos pelo `fetch`, quando utilizado o `server side`, são memorizados até que o componente seja renderizado, por padrão, mas podem ser ajustados passando mais parâmetros para a função.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+### revalidatePath
+
+A função revalidatePath é bem interessante quando alguma rota precisa de sofrer alguma atualização devido um evento.
```
```suggestion
A função `revalidatePath` é bem interessante quando alguma rota precisa sofrer alguma atualização devido a um evento.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+### Configurações de rota
+
+Uma rota estática pode ser forçada a ser dinâmica ou pode receber um tempo entre revalidações recebendo "atributos". Para forçar a página a ser dinâmica, é preciso colocar a linha `export const dynamic = "force-dynamic";` e para revalidar é possível adicionando `export const revalidate = 3600; // tempo em segundos`.
```
```suggestion
Uma rota estática pode ser forçada a ser dinâmica ou pode receber um tempo entre revalidações recebendo "atributos". Para forçar a página a ser dinâmica, é preciso colocar a linha `export const dynamic = "force-dynamic";` e para revalidar, é necessário adicionar `export const revalidate = 3600; // tempo em segundos`.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+### Client cache
+
+O next cria um `cache` de rotas no navegador do usuário. Se esse `cache` estiver atrapalhando alguma funcionalidade, poderá ser utilizada a função `router.refresh()` do hook `useRouter`.
```
Eu acho que cabe colocar um link pra documentação aqui.

```suggestion
O next cria um `cache` de rotas no navegador do usuário. Se esse `cache` estiver atrapalhando alguma funcionalidade, poderá ser utilizada a função [`router.refresh()` do hook `useRouter`](https://nextjs.org/docs/app/api-reference/functions/use-router#userouter).
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+## Optimizações
+
+O next apresenta optimizações `built-in` para melhorar a velocidade da aplicação e as [métricas da web](https://web.dev/articles/vitals?hl=pt-br). As `Images`, `Fonts` e `Metadata` serão abordadas nesse tópico, para mais recursos de optimização acesse [optimizações](https://nextjs.org/docs/app/building-your-application/optimizing).
```
```suggestion
## Otimizações

O next apresenta otimizações `built-in` para melhorar a velocidade da aplicação e as [métricas da web](https://web.dev/articles/vitals?hl=pt-br). As `Images`, `Fonts` e `Metadata` serão abordadas nesse tópico. Para mais recursos de otimização, acesse [*optimizing*](https://nextjs.org/docs/app/building-your-application/optimizing).
```

PedroAmorimG replied:
acho que aqui seria optimização mesmo porque seria uma parte dedicada a melhorias.

EdPPF replied:
Certeza? Otimização é justamente isso, não sei se optimização é certo (o corretor automático até reclama aqui)

PedroAmorimG replied:
Eu pesquisei optimização no google e tinha aparecido algo, mas eu vi aqui de novo e realmente parece que não existe vou trocar aqui.

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+### Imagens
+
+O next tem um componente chamado `Image` que extende a funcionalidade da `img` do html. Esse componente optimiza
```
```suggestion
O next tem um componente chamado `Image` que extende a funcionalidade da `img` do html. Esse componente otimiza
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+O next tem um componente chamado `Image` que extende a funcionalidade da `img` do html. Esse componente optimiza
+o tamanho do arquivo de imagem, promove estabilidade visual e utiliza `lazy loading`.
+
+O componente automaticamente converte a imagem para formatos modernos de imagem como `WebP` e `ÀVIF`.
```
```suggestion
O componente automaticamente converte a imagem para formatos modernos de imagem como `WebP` e `AVIF`.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+Se estiver usando tailwind no projeto, acesse [com TailWind](https://nextjs.org/docs/app/building-your-application/optimizing/fonts#with-tailwind-css)
+!!!
+
+As `google fonts` podem ser importadas do next e utilizadas sem nenhuma configuração em outra arquivo.
```
```suggestion
As `google fonts` podem ser importadas no next e utilizadas sem nenhuma configuração em outro arquivo.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+## Interações via Server Actions
+
+Outra forma interessante de realizar interações com o servidor é a `Server Action`, com ela é possível criar componentes que fazem chamadas ao servidor para realizar alguma ação. Essas ações podem ser mudanças no banco de dados com validação feita no servidor, por exemplo.
```
```suggestion
Outra forma interessante de realizar interações com o servidor é a `Server Action`. Com ela é possível criar componentes que fazem chamadas ao servidor para realizar alguma ação. Essas ações podem ser mudanças no banco de dados com validação feita no servidor, por exemplo.
```

EdPPF commented on stack/next.js/guia(app).md:
```diff
@@ -0,0 +1,903 @@
+
+Para criar uma `Server Action` é necessário inserir a diretiva `"use server"` no início de uma função assíncrona, nela será inserida a lógica que rodará no servidor. Aqui será explicada utilizando `Forms`, se quiser aprender utilizando `useEffect` ou `event handlers` entre [aqui](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#non-form-elements).
+
+Para conseguir acessar os dados do ´Forms´, deverá ser adicionado um parâmetro `formData` do tipo `FormData` na função que define a `server action`. Depois, na tag `<form action={ServerAction}>` do forms adicione a função criada no atributo `action`.
```
```suggestion
Para criar uma `Server Action`, é necessário inserir a diretiva `"use server"` no início de uma função assíncrona. Nela, será inserida a lógica que rodará no servidor. Aqui será explicada utilizando `Forms`. Se quiser aprender utilizando `useEffect` ou `event handlers` entre [aqui](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#non-form-elements).

Para conseguir acessar os dados do `Forms`, deverá ser adicionado um parâmetro `formData` do tipo `FormData` na função que define a `server action`. Depois, na tag `<form action={ServerAction}>` do forms, adicione a função criada no atributo `action`.
```

EdPPF reviewed (changes requested):
Maioria foi eu sendo meio chato com português de novo.

Mas eu acho que talvez você queira adicionar links que você acessou da documentação oficial do Next em cada seção daqui. Acho que isso ajuda o leitor a ir atrás das informações sozinho, caso ele precise. Dei um exemplo disso na parte de cache.

PedroAmorimG committed:
Correções feitas.

PedroAmorimG committed:
Correção otimização

PedroAmorimG committed:
Divisão em páginas, mudanças nos icones, nomes e ordem

marqueswill committed:
Changes to be committed:
	new file:   retype.yml
	new file:   stack/next-auth/callbacks.md
	renamed:    stack/next-auth/explicacao.md -> stack/next-auth/first-steps.md
	deleted:    stack/next-auth/implementacao.md
	new file:   stack/next-auth/providers.md
	new file:   stack/next-auth/routes.md
	new file:   stack/next-auth/session.md

marqueswill committed:
Changes to be committed:
	deleted:    stack/next-auth/index.yml

marqueswill committed:
Changes to be committed:
	new file:   stack/next-auth/index.yml
	deleted:    stack/retype.yml

marqueswill committed:
Changes to be committed:
	deleted:    retype.yml
	new file:   stack/retype.yml

pafev committed:
Merge branch 'main' into feat-AppNextjs

EdPPF commented on stack/next.js/iniciando.md:
```diff
@@ -14,9 +14,10 @@ Next não é instalado de fato em sua máquina, é um conjunto de pacotes que s
+
+-   É necessário possuir uma versão igual ou superio a 16.14 do Node.js.
+-   São suportados macOS, Windows (incluindo WSL) e Linux.
+    !!!
```
Acho que a indentação desse `!!!` tá quebrando a visualização dessa página. Melhor dar uma revisada nessas exclamações desse arquivo, acho que tem algo errado nele.
Aparece assim pra mim:
![docStructError](https://github.com/StructCE/DocStruct/assets/90335387/f2ae96da-a3cd-43d1-b4fb-fd1634fa4c03)
E vai até o final:
![docStructError2](https://github.com/StructCE/DocStruct/assets/90335387/e0d2658a-9e64-47f1-b2d7-730a241e613c)

PedroAmorimG committed:
Correção

pafev commented on stack/next.js/explicacao.md:
vc e o Matheus querem assustar a glr? que tanto de sigla e termo em inglês é esse? seria bom tornar a introdução mais amigável a trainee e a quem tá conhecendo o Next

pafev commented on stack/next.js/iniciando.md:
```diff
@@ -38,11 +40,13 @@ Would you like to use App Router? (recommended) No / Yes
+3. Uso do ESLint, um analisador estático de código, que encontra problemas no código e alerta no editor de texto.
+4. Uso do Tailwind CSS, a principal ferramenta de estilização usada em Next.js.
 5. Opção do diretório `src/` na pasta raiz do projeto, mais uma opção de organização para separar a pasta do projeto dos arquivos de configuração.
 6. Opção de tipo de roteamento usado Pages Router e App Router (mais recente), nessa documentação iremos usar o Pages Router. Caso queira saber mais sobre a diferença entre os dois
```
O foco da documentação não é mais pages router

pafev commented on stack/next.js/iniciando.md:
```diff
@@ -59,26 +63,26 @@ Em seguida adicione o trecho de código ao arquivo `package.json`:
 Como a configuração, neste caso, é manual, então algumas pastas não vem pré-configuradas no projeto. Logo, deve-se criar a pasta pages (pois estamos usando pages router) onde ficará as páginas a serem usadas na aplicação. Assim como, pode-se criar a pasta public, para arquivos estáticos como imagens, fontes, ícones etc.
 
 !!!
 Mais informações sobre como iniciar e configurar um projeto Next [aqui](https://nextjs.org/docs/getting-started/installation#the-pages-directory-optional).
```
o foco da documentação não é mais pages router (não somos ultrapassados aiai)

pafev commented on stack/next.js/iniciando.md:
```diff
@@ -59,26 +63,26 @@ Em seguida adicione o trecho de código ao arquivo `package.json`:
 Mais informações sobre como iniciar e configurar um projeto Next [aqui](https://nextjs.org/docs/getting-started/installation#the-pages-directory-optional).
-!!!
\ No newline at end of file
+!!!
```
pergunta sincera: qual a utilidade da configuração manual?

pafev commented on stack/next.js/intro_roteamento.md:
```diff
@@ -0,0 +1,258 @@
+
+O `App Router` foi introduzido na versão 13 do next.js e foi construído utilizando o `React Server Components (RSCs)`.
+
+O `App Router` trabalha com o diretório app para a criação de rotas que podem ou não ser de acesso público. Por padrão, componentes deste diretório são `RSCs`. As páginas são criadas nas rotas e irão herdar certas características definidas por arquivos especiais, como `navbar` e `footer`, definidos no arquivo `layout.tsx` da rota.
```
```suggestion
O `App Router` trabalha com o diretório app para a criação de rotas que podem ou não ser de acesso público. Por padrão, componentes deste diretório são `RSCs`.
```
vc já vai falar de rotas e layout logo abaixo, acho que não precisa adiantar isso e aumentar o texto

pafev commented on stack/next.js/intro_roteamento.md:
ficou mt bom

pafev commented on stack/next.js/api_serverAction.md:
```diff
@@ -0,0 +1,350 @@
+
+	const excluirMensagem = async (id: number) => {
+		"use server";
+	};
```
esqueceu de apagar ou era pra ter código aq? kkk

pafev commented on stack/next.js/guia(pages).md:
```diff
@@ -1,7 +1,7 @@
 order: 1
 icon: book
-label: "Guia prático"
+label: "Guia prático (Pages Router)"
```
```suggestion
label: "Pages Router"
```

acho que esse label faz mais sentido

pafev commented on stack/next.js/render_cache_opt.md:
```diff
@@ -0,0 +1,315 @@
+
+### Configurações de rota
+
+Uma rota estática pode ser forçada a ser dinâmica ou pode receber um tempo entre revalidações recebendo "atributos". Para forçar a página a ser dinâmica, é preciso colocar a linha `export const dynamic = "force-dynamic";` e para revalidar, é necessário adicionar `export const revalidate = 3600; // tempo em segundos`.
```
um detalhe do `export cons dynamic` é que ele só pode ser usado em um arquivo page.tsx, mas se vc quer passar isso em um componente, por modularização e divisão das responsabilidades, pode utilizar o `unstable_noStore as noStore` do `next/cache`. Eles colocam esse unstable, mas nunca vi dar problema

se quiser falar sobre... se não, tudo bem

pafev commented on stack/next.js/render_cache_opt.md:
```diff
@@ -0,0 +1,315 @@
+
+### Client cache
+
+O next cria um `cache` de rotas no navegador do usuário. Se esse `cache` estiver atrapalhando alguma funcionalidade, poderá ser utilizada a função [`router.refresh()` do hook `useRouter`](https://nextjs.org/docs/app/api-reference/functions/use-router#userouter).
```
ainda acho melhor usar uma server action que tenha só revalidatePath("/path", "page"), como
```
async function refresh() {
  "use server";
  revalidatePath("/path", "page");
}
```
pq ela tbm faz uma revalidação inteligente da rota, sem dar um reload da página, e não precisa forçar um componente a ser client só pra usar o `router.refresh()`
obs.: é server action até para poder ser passada como props de um client component, se quiser reaproveitar em um client component

PedroAmorimG replied:
não lembro o q eu tava pensando na epoca mas pra mim nao faz mais sentido kk 
esse cache existiria provavelmente por erro na manipulação de cache e a solução certa talvez não seja criar um meio para o usuario sinalizar que precisa de uma atualização ao servidor ou a propria pagina pedir um refresh.

pafev reviewed (changes requested):
a doc ficou boa. Sucinta e prática

PedroAmorimG committed:
alterações pedidas

pafev committed:
fix pages to app  router recommendation (iniciando.md)

pafev committed:
Merge branch 'main' into feat-AppNextjs
